build-tools/timing/timing.targets

```diff
@@ -0,0 +1,153 @@
+  <Import Project="timing.projitems" />
+  <Import Project="..\scripts\TestApks.targets" />
+  <UsingTask AssemblyFile="..\..\bin\Build$(Configuration)\xa-prep-tasks.dll" TaskName="Xamarin.Android.BuildTools.PrepTasks.Git" />
+  <UsingTask AssemblyFile="..\..\bin\Build$(Configuration)\xa-prep-tasks.dll" TaskName="Xamarin.Android.BuildTools.PrepTasks.GitCommitHash" />
+  <UsingTask AssemblyFile="..\..\bin\Build$(Configuration)\xa-prep-tasks.dll" TaskName="Xamarin.Android.BuildTools.PrepTasks.ProcessMSBuildTiming" />
+  <Target Name="LogcatTiming">
+    <Exec Command="&quot;$(AdbToolPath)\$(AdbToolExe)&quot; $(_AdbTarget) $(AdbOptions) logcat -v threadtime -d > tmp-logcat-timing.log" />
+    <ProcessLogcatTiming
+        InputFilename="tmp-logcat-timing.log"
+        ApplicationPackageName="$(Package)"
+        PID="$(PID)"
+        ResultsFilename="tmp-timing-results.csv"
+        DefinitionsFilename="..\scripts\TimingDefinitions.txt"
+        AddResults="true"
+        Activity="$(Activity)" />
+  </Target>
+  <PropertyGroup>
+    <_MSBuildTimingDependsOn>
+      MSBuildPrep;
+      FreshBuild;
+      FreshInstall;
+      SecondBuild;
+      SecondInstall;
+      TouchCSharpBuild;
+      TouchCSharpInstall;
+      TouchAndroidResourceBuild;
+      TouchAndroidResourceInstall;
+    </_MSBuildTimingDependsOn>
+  </PropertyGroup>
+  <Target Name="MSBuildTiming" DependsOnTargets="$(_MSBuildTimingDependsOn)">
+    <ItemGroup>
+      <_TimingResults Include="$(_OutputDir)Timing_*.xml" />
+    </ItemGroup>
+    <ProcessMSBuildTiming
+        InputFiles="@(_TimingResults)"
+        ResultsFilename="$(_TopDir)TestResult-MSBuild-times.csv"
+        LabelSuffix="-$(Configuration)"
+    />
+  </Target>
+  <Target Name="MSBuildPrep" DependsOnTargets="ResolveReferences;AcquireAndroidTarget">
+    <PropertyGroup>
+      <_OutputDir>$(_TopDir)bin\Test$(Configuration)\</_OutputDir>
+      <_XABuild>$(_TopDir)bin\Debug\bin\xabuild</_XABuild>
+      <_TimingLogger>Xamarin.Android.Timing.TimingLogger,$(_OutputDir)Xamarin.Android.Timing.dll</_TimingLogger>
+    </PropertyGroup>
+    <ItemGroup>
+      <_FilesToDelete Include="$(_OutputDir)Timing_*" />
+    </ItemGroup>
+    <Delete Files="@(_FilesToDelete)"/>
+    <GitCommitHash
+        WorkingDirectory="$(_TopDir)"
+        ToolPath="$(GitToolPath)"
+        ToolExe="$(GitToolExe)">
+      <Output TaskParameter="AbbreviatedCommitHash" PropertyName="_XAHash" />
+    </GitCommitHash>
+  </Target>
+  <Target Name="FreshBuild"
+      Outputs="$(_TopDir)bin\Test$(Configuration)\Timing_FreshBuild_$(Configuration)_%(XamarinAndroidProject.Name).xml">
+    <PropertyGroup>
+      <_ID>%(XamarinAndroidProject.ShortName)-FreshBuild</_ID>
+      <_Description>A fresh build after a clean checkout for %(XamarinAndroidProject.Name)</_Description>
+      <_Target>SignAndroidPackage</_Target>
+      <_ProjectDir>$([System.IO.Path]::GetDirectoryName(%(XamarinAndroidProject.Identity)))</_ProjectDir>
```

Comment:
`xbuild` is calculating `$(_ProjectDir)` as an empty string.

Is there some workaround I need here? It works for `msbuild`, even on macOS.

Reply:
I don't see how it could be evaluated as non-empty on msbuild. `%(XamarinAndroidProject.Identity)` is evaluated to "%(XamarinAndroidProject.Identity)" (as is). You'd need `@(XamarinAndroidProject->'%(Identity)')`. It equally failed on both msbuild and xbuild.

```<Project xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
  <ItemGroup>
    <XamarinAndroidProject Include="/sources/xamarin-android/Xamarin.Android.sln" />
  </ItemGroup>
  <PropertyGroup>
    <_V>%(XamarinAndroidProject.Identity)</_V>
    <_W>@(XamarinAndroidProject->'%(Identity)')</_W>
  </PropertyGroup>
  <Target Name="Build">
    <Message Text="V: $(_V)" />
    <Message Text="W: $(_W)" />
  </Target>
</Project>

$ msbuild x.targets 
</snip>
  V: %(XamarinAndroidProject.Identity)
  W: /sources/xamarin-android/Xamarin.Android.sln
```

I still couldn't find a valid solution, but you might find it with this hint.

Reply:
Yeah it must be an `xbuild` bug, if you make the target "batch" by adding an `Outputs` it works:
```xml
<Project xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
  <ItemGroup>
    <XamarinAndroidProject Include="/sources/xamarin-android/Xamarin.Android.sln" />
  </ItemGroup>
  <Target Name="Build" Outputs="%(XamarinAndroidProject.Identity)">
    <PropertyGroup>
      <_V>%(XamarinAndroidProject.Identity)</_V>
      <_W>@(XamarinAndroidProject->'%(Identity)')</_W>
    </PropertyGroup>
    <Message Text="V: $(_V)" />
    <Message Text="W: $(_W)" />
  </Target>
</Project>
```

msbuild:
```
V: /sources/xamarin-android/Xamarin.Android.sln
W: /sources/xamarin-android/Xamarin.Android.sln
```

xbuild:
```
V: %(XamarinAndroidProject.Identity)
W: /sources/xamarin-android/Xamarin.Android.sln
```

Anyway, thanks for your example, I see how to fix it now.